File: src/runtime/javascript/engine/nodes/support/table_column_pipeline.ts
import { Table } from "../../table";
import { columnMath, columnTransformSpec } from "./table_column_math";
import { requireTable } from "./common";

export function columnTransformOutput(params: Record<string, unknown>): Record<string, unknown> {
  return columnTransformSpec(params);
}

export function columnPipeline(upstream: unknown, params: Record<string, unknown>): Table {
  if (!upstream || typeof upstream !== "object" || Array.isArray(upstream) || upstream instanceof Table) throw new Error("Column transform Pipeline requires named inputs");
  const inputs = upstream as Record<string, unknown>;
  let value = requireTable(inputs.input, "Column transform Pipeline");
  const expected = Number(params.transformCount ?? 1);
  if (!Number.isInteger(expected) || expected < 1 || expected > 16) throw new Error("Column transform Pipeline transformCount must be between 1 and 16");
  const transforms = Object.entries(inputs).filter(([port]) => port.startsWith("transform")).map(([port, item]) => {
    const order = Number(port.slice(9));
    if (!Number.isInteger(order) || order < 1) throw new Error(`Invalid Column transform Pipeline port: ${port}`);
    if (!item || typeof item !== "object" || Array.isArray(item) || item instanceof Table) throw new Error(`Column transform Pipeline input ${port} must be a Transform object`);
    return [order, item as Record<string, unknown>] as const;
  }).sort((left, right) => left[0] - right[0]);
  if (transforms.length !== expected) throw new Error(`Column transform Pipeline requires ${expected} connected Transform inputs`);
  transforms.forEach(([order], index) => { if (order !== index + 1) throw new Error("Column transform Pipeline inputs must use consecutive Transform ports"); });
  for (const [, transform] of transforms) value = columnMath(value, transform);
  return value;
}

File: python/pydroid_flow/engine_parts/table_column_pipeline.py
from __future__ import annotations

from typing import Any

import pandas as pd

from .table_column_math import column_math, column_transform_spec
from .values import _require_table


def column_transform_output(params: dict[str, Any]) -> dict[str, Any]:
    return column_transform_spec(params)


def conditional_transform_output(upstream: Any, params: dict[str, Any]) -> dict[str, Any]:
    if not isinstance(upstream, dict) or not isinstance(upstream.get("transform"), dict):
        raise ValueError("Conditional transform requires a Transform input")
    return {**upstream["transform"], "enabled": bool(params.get("condition", True))}


def column_pipeline(upstream: Any, params: dict[str, Any]) -> pd.DataFrame:
    if not isinstance(upstream, dict):
        raise ValueError("Column transform Pipeline requires named inputs")
    value = _require_table(upstream.get("input"), "Column transform Pipeline")
    expected = int(params.get("transformCount", 1))
    if expected < 1 or expected > 16:
        raise ValueError("Column transform Pipeline transformCount must be between 1 and 16")
    transforms: list[tuple[int, dict[str, Any]]] = []
    for port, item in upstream.items():
        if not str(port).startswith("transform"):
            continue
        try:
            order = int(str(port)[9:])
        except ValueError as exception:
            raise ValueError(f"Invalid Column transform Pipeline port: {port}") from exception
        if not isinstance(item, dict):
            raise ValueError(f"Column transform Pipeline input {port} must be a Transform object")
        transforms.append((order, item))
    transforms.sort(key=lambda item: item[0])
    if len(transforms) != expected:
        raise ValueError(f"Column transform Pipeline requires {expected} connected Transform inputs")
    for expected_order, (actual_order, transform) in enumerate(transforms, start=1):
        if actual_order != expected_order:
            raise ValueError("Column transform Pipeline inputs must use consecutive Transform ports")
        if transform.get("enabled", True) is False:
            continue
        value = column_math(value, transform)
    return value
